### ISSUE TITLE
Quantum Ant - Developing a Quantum Calculator

### ISSUE THREAD

**Issue by ShuRaAtum** [qiskit-community/quantum-hackathon-korea-22#9]
# Abstract
Our quantum calculator includes a quantum adder implementation, the goal of the IBM challenge. Three quantum adders are available: ripple carry adder, improved ripple carry adder (CDKM adder), and carry lookahead adder. Furthermore, we implement arithmetic on binary fields (such as multiplication, squaring and inversion). Among operations in binary fields, the multiplication is essential and cost-critical even in the quantum field. In this hackathon, we optimize quantum binary field multiplication. We try to reduce the error of a 4-bit quantum adder. Lastly, we develop the UI (User Interface) for the quantum calculator. It includes all the arithmetic we implemented and is pretty cool. 

# Members
@amdjd0704
@ShuRaAtum
@kyungzzu
@YangYu34
@starj1023

# Introduction
Quantum algorithms to solve practical problems in quantum chemistry, materials science, and matrix inversion often involve a significant amount of arithmetic operations which act on a superposition of inputs. These have to be compiled to a set of fault-tolerant low-level operations and throughout this translation process, the compiler aims to come close to the Pareto-optimal front between the number of required qubits and the depth of the resulting circuit. 
Quantum computers capable of running the Shor algorithm can solve discrete logarithm problems, which are the challenges of Ellip- tic Curve Cryptography (ECC). As such, it is well known that the Shor’s algorithm violates the security of public key cryptography. An attacker equipped with a large-scale quantum computer would undermine the safety of the ECC. In this current situation, the analysis of potential quantum computer attacks on public key cryptography should be considered for a robust security system. Therefore, it is probably a commendable plan to consider the quantum security of the public-key ciphers, to be on the safe side (See https://eprint.iacr.org/2020/1296.pdf). 
One of the major overheads in applying Shor’s algorithm is implementing binary elliptic curve arithmetic in quantum circuits. 

Keeping this in mind, in this hackathon, we present a quantum calculator with various quantum additions and quantum binary field arithmetic.

# Quantum Addition

1. **Ripple-carry Adder**
A generic quantum ripple-carry adder is implemented as a combination of MAJ and UMA composed of one Toffoli gate and two CNOT gates (Figure 1). When adding a and b, the result value is stored in b, and two additional qubits are used: c0 to calculate the initial carry value and z to store the highest carry value.
<img width="550" alt="스크린샷 2022-06-29 오전 11 09 07" src="https://user-images.githubusercontent.com/55376144/176336096-9d2f45f0-179d-4964-a9ce-af3a038333c4.png">

2.  **Improved Ripple-carry Adder(n>=4) (CDKM)**
This quantum ripple-carry adder enables an improved implementation of the quantum adder when the addition unit n is greater than or equal to 4. Since the minimum unit of addition n of this Challenge Topic is 4, we implement an improved ripple-carry adder and compare it with the ripple-carry adder (which is generic). Figure 2 shows the quantum circuit structure of the improved ripple carry adder.
<img width="378" alt="스크린샷 2022-06-29 오전 11 09 20" src="https://user-images.githubusercontent.com/55376144/176336152-513bd09f-412c-41cc-aeec-2db1dbc05486.png">

3.  **Quantum Carry-lookahead Adder (QCLA)**
The carry-lookahead adder provides the fastest of all quantum adders by using a method that computes bits in parallel. The adder is divided into an out-of-place method and an in-place method (result is the same). In the out of place method, a total of ???? ancilla qubits are used in the addition of a and b of n-bit length. The left side of Figure 3 shows a quantum addition circuit with an out-of-place approach for a 10-bit input.
<img width="579" alt="스크린샷 2022-06-29 오전 11 09 36" src="https://user-images.githubusercontent.com/55376144/176336183-d985699b-cc23-48c8-a836-c7fa34bc00dc.png">

# Binary Field Arithmetic
**Quantum Multiplication**
Previous works on quantum multiplication focused on minimizing the number of Toffoli gates or qubits. In contrast, our work presents strate- gies for optimizing Toffoli depth and full depth, which are key factors in the Noisy Intermediate-Scale Quantum (NISQ) era. To achieve our goal, Karatsuba multiplication using divide-and-conquer approach is adopted. In a nutshell, we present an optimized quantum multiplication with Tof- foli depth one. Furthermore, under the influence of the optimized Toffoli depth, the full depth is naturally reduced. In order to show the effectiveness of the proposed method, the performance is evaluated by various metrics, such as, qubits, quantum gates, depth, and qubits-depth product.

1. **Schoolbook Multiplication**
Multiplication of GF(2^n) performs n-bit polynomial multiplication and modular reduction by irreducible polynomial N . The multiplication of f and g of GF(2^n).

For the generated product of f and g, the modular reduction is performed over n bits in length. As a result, h, the product of f and g, becomes an element of GF(2^n). As illustrated in Figure 5, f · g with Schoolbook multiplication (which is general) requires n^2 Toffoli gates.

2. **Karatsuba Multiplication**
It is well known that the Karatsuba algorithm can reduce the complexity of multiplication by using addition operations. For the multiplication of polynomials f and g of size n (i.e, h = f · g), the Karatsuba algorithm divides the two input polynomials f and g into the size of s = n/2

3. **Our Karatsuba Multiplication**
In the proposed method, we apply the Karatsuba algorithm once to reduce the size of the multiplication, which is called Level-1. In Level-1, it is divided into three multiplications f0 · g0, f1 · g1, and (f0 + f1) · (g0 + g1). The size of each multiplication is reduced to (n/2) and 3 · (n/2)^2 Toffoli gates are required. This result reduces the use of Toffoli gates from n^2 to 3·(n/2)^2, but may not be completely parallel, which means the three multiplications are not performed simultaneously. In the Level-1 layer of Figure 8, multiplications f0 · g0 (low part) and f1 · g1 (high part) are performed, simultaneously, but (f0 + f1) · (g0 + g1) in the rectangle (middle part) is performed sequentially after the previous multiplications are finished. Operands of the multiplication of the middle part (f0 +f1, g0 +g1) are overwritten in f0, g0 or f1, g1 (i.e., f0 = f0 +f1, g0 = g0 +g1), and it is possible only after the multiplication of the high part and the low part is finished. Before multiplications, we allocate clean qubits (rectangle of Level-1 layer in Figure 8) and independently prepare the middle part operands (f0 + f1) and (g0 + g1) using CNOT gates. Preparing this middle part requires n clean qubits and 2n CNOT gates. (i.e. 2/n clean qubits = f0 + f1 and 2/n clean qubits = g0 + g1) Three multiplications of low, middle, and high become independent of each other and can be performed, simultaneously. As a result, in Level-1, quantum multiplication of reduced Toffoli depth is presented. If the Karatsuba multiplication is terminated, additions between the generated products f0 · g0, f1 · g1, and (f0 + f1) · (g0 + g1) are performed using CNOT gates to complete the multiplication. Furthermore, we apply the Karatsuba algorithm recursively to optimize the Toffoli depth and full depth.
In Level-2, the Karatsuba algorithm is applied to each of the three multiplications divided by Level-1. Similar to Level-1, in Level-2 there is a dependency on multiplications for middle parts. Thus, we allocate 3 × clean qubits and prepare the 3 × middle parts using CNOT gates again (rectangles of Level-2 layer in Figure 8). Preparing these middle parts requires 3 · (n/2) clean qubits and 3n CNOT gates. As a result, nine multiplications become completely independent of each other and are performed simultaneously. In Level-2, Toffoli gates and Toffoli depth that were reduced in Level-1 are reduced once more to 3^2 · (n/2^2 )2 for Toffoli gates and (3n/2^2 − 2) for Toffoli depth. In this way, the Karatsuba algorithm is applied recursively until it is divided into multiplications of size one, and the dependencies of all multiplications are released. Then, we can perform quantum multiplication of Toffoli depth one by generating all products in parallel. Finally, the proposed quantum multiplica- tion circuit achieves the best performance of Toffoli depth one, providing high performance even for full depth.

In quantum multiplication, Toffoli depth has a huge impact on counting full depth. Thus, under the influence of the optimized Toffoli depth, our work achieves the lowest full depth. It is hard to say that quantum computers in the upcoming NISQ era are small anymore (Figure 9), but they need quantum error correction. The metric of Toffoli depth is probably the most important. However, since the number of qubits is still important, it is important to consider the trade-off between qubits and depth. 

Our quantum multiplication requires relatively more qubits, but achieves the best trade-off of (Toffoli depth ✕ Qubits). This metric represents the trade-off for quantum circuits and is adopted in Asiacrypt paper (https://link.springer.com/chapter/10.1007/978-3-030-64834-3_24).

# Quantum Squaring 
The squaring operation in a binary field is simple to implement. Figure 10 shows the squaring quantum circuit structure for n=10. Since general squaring operation is implemented only by left shift operation between Qubits, quantum resources are not used by changing the index of Qubits. Therefore, assuming that the squaring operation is performed, the modular operation for n > 10 is implemented using CNOT gates.
<img width="236" alt="스크린샷 2022-06-29 오전 11 09 54" src="https://user-images.githubusercontent.com/55376144/176336247-fa27a4d2-ed34-42cb-aee6-ca55e50cae30.png">

# Quantum Inversion
In the case of a general non-finite field inversion operation, it can be easily computed. In contrast, the inversion operation in a finite field has the highest computational complexity. A representative way to implement the inversion operation in a finite field is to implement the Itoh-Tsuji inversion operation algorithm. ​​The Itoh-Tsuji inversion calculation algorithm is implemented as a combination of multiplication and square operations. Therefore, if the square and multiplication operation, which is the implementation goal above, is implemented, the inversion operation can be implemented through an appropriate combination according to the defined finite field. Since the squaring operation is relatively simple, applying our optimized Karatsuba multiplication also optimizes the inversion.

# UI for Quantum Calculator
We made a cool quantum calculator. There are two versions, one running on the simulator and the other running on a fake backend. This calculator calculates using several implementations of Quantum Adder and Quantum Binary Field Arithmetic. Usage is similar to a classic calculator. In addition, the count result of circuit execution is output as a histogram on the right.
<img width="1298" alt="calculator" src="https://user-images.githubusercontent.com/55376144/176335786-68011f02-cea1-4dca-bede-ae62c36f82e3.png">

# Quantum Error Correction
Coding theory for correcting noise generated in communication channels a few decades ago is similarly used to correct errors that occur in current quantum computers. As shown in Figure 12, the target qubit is encoded using additional qubits before noise occurs, and the error of the target qubit that has an error is corrected after operation. Finally, decoding is performed to obtain error-free qubit values.

# Conclusion
In response to the quantum era, quantum arithmetic optimization is interesting to the quantum community. We present efficient quantum calculators essential for quantum computing such as QML(Quantum Machine Learning) or quantum simulations and quantum cryptanalysis on ECC. 
However, we failed to completely correct the error. So our future work is to completely correct errors in quantum calculators. 

# Appendix
This section shows the result of the real Backend running our implementation. 
First, Fig 12. shows the result of the Ripple-carry Adder. The input values are 0111, 1001, and the correct answer is 1000. However, in the real backend, 00000 is the highest probability.
<img width="378" alt="스크린샷 2022-06-29 오전 11 13 24" src="https://user-images.githubusercontent.com/55376144/176336372-83007bb6-17d9-4d1c-8b24-35ed35312973.png">

Second, Fig 13. shows the result of the Improved Ripple-carry Adder(CDKM). The input values are 1001, 1000, and the correct answer is 10000. However, in the real backend, 00101 is the highest probability.
<img width="483" alt="스크린샷 2022-06-29 오전 11 13 31" src="https://user-images.githubusercontent.com/55376144/176336406-73033ae5-2f53-4bf8-82a2-14236ac8e8e8.png">

Third, Fig 14. shows the result of the Karatsuba Multiplecation. The input values are 1001, 0111, and the correct answer is 1101. However, in the real backend, 0000, 1100 are the highest probability.
<img width="470" alt="스크린샷 2022-06-29 오전 11 13 38" src="https://user-images.githubusercontent.com/55376144/176336411-d8efa4eb-dfb3-40be-9f7d-cca2aa9c2bcc.png">

# Reference
[Ripple-carry] 
S. A. Cuccaro et al. “A new quantum ripple-carry addition circuit.”, (https://arxiv.org/abs/quan
tph/0410184)
[QFT Adder]
L. Ruiz-Perez et al. & Quantum arithmetic with the Quantum Fourier Transform.&, (https://arxiv.org/abs/1411.5949)
[Schoolbook MUL]
D. Maslov et al. “ON THE DESIGN AND OPTIMIZATION OF A QUANTUM POLYNOMIAL-TIME ATTACK ON ELLIPTIC CURVE CRYPTOGRAPHY.”
[Karatsuba MUL 1]
S. Kepley et al. “Quantum circuits for F2n-multiplication with subquadratic gate count.”
[Karatsuba MUL 2]
V. Hoof. “Space-efficient quantum multiplication of polynomials for binary finite fields with sub-quadratic Toffoli gate count”
[Quantum Error Correction]
https://qiskit.org/textbook/ch-quantum-hardware/error-correction-repetition-code.html

# GitHub repo
https://github.com/ShuRaAtum/2022Qhackathon_QuantumAnt
